packages/block-editor/src/components/provider/index.js

```diff
@@ -14,21 +14,81 @@ import { BlockRefsProvider } from './block-refs-provider';
 
 /** @typedef {import('@wordpress/data').WPDataRegistry} WPDataRegistry */
 
-function BlockEditorProvider( props ) {
-	const { children, settings } = props;
+export const ExperimentalBlockEditorProvider = withRegistryProvider(
+	( props ) => {
+		const { children, settings } = props;
 
-	const { updateSettings } = useDispatch( blockEditorStore );
-	useEffect( () => {
-		updateSettings( {
-			...settings,
-			__internalIsInitialized: true,
-		} );
-	}, [ settings ] );
+		const { updateSettings } = useDispatch( blockEditorStore );
+		useEffect( () => {
+			updateSettings( {
+				...settings,
+				__internalIsInitialized: true,
+			} );
+		}, [ settings ] );
 
-	// Syncs the entity provider with changes in the block-editor store.
-	useBlockSync( props );
+		// Syncs the entity provider with changes in the block-editor store.
+		useBlockSync( props );
 
-	return <BlockRefsProvider>{ children }</BlockRefsProvider>;
-}
+		return <BlockRefsProvider>{ children }</BlockRefsProvider>;
+	}
+);
 
-export default withRegistryProvider( BlockEditorProvider );
+/**
+ * The public BlockEditorProvider does not accept any experimental settings
+ * introduced after WordPress 6.1. To use them, you need to rely on
+ * ExperimentalBlockEditorProvider.
+ *
+ * @see https://github.com/WordPress/gutenberg/pull/46131
+ */
+const publicExperimentalSettingsFromBeforeWordPress62 = [
+	'__experimentalBlockDirectory',
+	'__experimentalDiscussionSettings',
+	'__experimentalFeatures',
+	'__experimentalPreferredStyleVariations',
+	'__experimentalSetIsInserterOpened',
+	'__unstableGalleryWithImageBlocks',
+	'__experimentalReusableBlocks',
+	'__experimentalBlockPatterns',
+	'__experimentalBlockPatternCategories',
+	'__experimentalAdditionalBlockPatterns',
+	'__experimentalAdditionalBlockPatternCategories',
+	'__experimentalFetchLinkSuggestions',
+	'__unstableEnableFullSiteEditingBlocks',
+	'__unstableIsBlockBasedTheme',
+	'__unstableFetchMedia',
+	'__experimentalFetchUrlData',
+	'__experimentalFetchRichUrlData',
+	'__experimentalCanUserUseUnfilteredHTML',
+	'__experimentalUndo',
+	'__experimentalCreatePageEntity',
+	'__experimentalUserCanCreatePages',
+	'__experimentalPreferPatternsOnRoot',
+	'__unstableIsPreviewMode',
+];
+const BlockEditorProvider = ( props ) => {
+	const cleanSettings = useMemo( () => {
+		const clean = {};
+		for ( const key in props.settings ) {
+			if (
+				key.startsWith( '__' ) &&
```

Comment:
I think we decided to not use the `__experimental` prefix for experimental props on a `lock()`ed component, yeah? So this check isn't sufficient as it won't remove experimental props with no `__` prefix.

Reply:
In this case it's not merely about the component props, but about the block editor settings. I don't think the same logic applies. What do you think? If we omit the `__experimental` prefix on those then this check will need to go check the incoming props against through an exhaustice allow/denylist of experimental properties.

Reply:
I'm not sure 😅 just trying to be consistent. It would be confusing to know which settings are experimental vs. which ones aren't if we do not have the `__experimental` prefix.

At any rate since they will be private we can change our minds about this later.

Reply:
There is a filter:

```php
add_filter( 'block_editor_settings_all', 'gutenberg_get_block_editor_settings', PHP_INT_MAX );
```

What's more, there is no schema to enforce a specific shape of the settings object. You can add any settings you want to. An allowlist won't work here, then. Too bad.

It's down to the `__` prefix and a risky `aListOfAllExperimentalProperties`. I'll move forward with the first one in this PR, and if that turns out to be a problem then we can always revisit (as you mentioned).

Reply:
What about only having a list with the settings we consider experimental. I know it's a bit error-prone as someone might forget to add it here, but doesn't rely on any naming convention and extenders could actually use whatever they want.

I'll think about it a bit more though :)

Reply:
It would be good to add this to the docs, that we need to also add it this array.